networking/server.py

```diff
@@ -0,0 +1,133 @@
+import asyncio
+import json
+import pickle
+import base64
+import websockets
+import time
+import traceback
+from pathlib import Path
+from datetime import datetime
+from robotouille.robotouille_env import create_robotouille_env
+
+SIMULATE_LATENCY = False
+SIMULATED_LATENCY_DURATION = 0.25
+
+def run_server(environment_name: str, seed: int=42, noisy_randomization: bool=False, display_server: bool=False, event: asyncio.Event=None):
+    asyncio.run(server_loop(environment_name, seed, noisy_randomization, display_server))
+
+async def server_loop(environment_name: str, seed: int=42, noisy_randomization: bool=False, display_server: bool=False, event: asyncio.Event=None):
+    waiting_queue = {}
+    reference_env = create_robotouille_env(environment_name, seed, noisy_randomization)[0]
+    num_players = len(reference_env.get_state().get_players())
+
+    async def simulator(connections):
+        try:
+            print("Start game", connections.keys())
+            recording = {}
+            recording["start_time"] = datetime.now().strftime("%Y%m%d_%H%M%S_%f")
+            recording["environment_name"] = environment_name
+            recording["seed"] = seed
+            recording["noisy_randomization"] = noisy_randomization
+            recording["actions"] = []
+            recording["violations"] = []
+            start_time = time.monotonic()
+
+            env, json_data, renderer = create_robotouille_env(environment_name, seed, noisy_randomization)
+            obs, info = env.reset()
+            if display_server:
+                renderer.render(obs, mode='human')
+            done = False
+            interactive = False  # Adjust based on client commands later if needed
+
+            assert len(connections) == num_players
+            sockets_to_playerID = {}
+            for i, socket in enumerate(connections.keys()):
+                sockets_to_playerID[socket] = i
+                player_data = base64.b64encode(pickle.dumps(i)).decode('utf-8')
+                opening_message = json.dumps({"player": player_data})
+                await socket.send(opening_message)
+
+            while not done:
+                # Wait for messages from any client
+                # TODO(aac77): 
```

Comment:
make a Git issue and refer to it in the PR as well in the code by the number, e.g. #32